src/libs/Network/PersistedRequestsQueue.js

```diff
@@ -0,0 +1,96 @@
+ *
+ * @returns {Promise}
+ */
+function process() {
+    const persistedRequests = PersistedRequests.getAll();
+
+    // This sanity check is also a recursion exit point
+    if (NetworkStore.getIsOffline() || _.isEmpty(persistedRequests)) {
+        return Promise.resolve();
+    }
+
+    const tasks = _.map(persistedRequests, request => processRequest(request)
+        .then((response) => {
+            if (response.jsonCode === CONST.JSON_CODE.NOT_AUTHENTICATED) {
+                NetworkEvents.getLogger().info('Persisted optimistic request needs authentication');
+            } else {
+                NetworkEvents.getLogger().info('Persisted optimistic request returned a valid jsonCode. Not retrying.');
+            }
+            NetworkEvents.onResponse(request, response);
+            PersistedRequests.remove(request);
+        })
+        .catch((error) => {
+            // If we are catching a known network error like "Failed to fetch" allow this request to be retried if we have retries left
+            if (error.message === CONST.ERROR.FAILED_TO_FETCH) {
+                const retryCount = PersistedRequests.incrementRetries(request);
+                NetworkEvents.getLogger().info('Persisted request failed', false, {retryCount, command: request.command, error: error.message});
+                if (retryCount >= CONST.NETWORK.MAX_REQUEST_RETRIES) {
+                    NetworkEvents.getLogger().info('Request failed too many times removing from storage', false, {retryCount, command: request.command, error: error.message});
+                    PersistedRequests.remove(request);
+                }
+            } else if (error.name === CONST.ERROR.REQUEST_CANCELLED) {
+                NetworkEvents.getLogger().info('Persisted request was cancelled. Not retrying.');
+                NetworkEvents.onError(request);
+                PersistedRequests.remove(request);
+            } else {
+                NetworkEvents.getLogger().alert(`${CONST.ERROR.ENSURE_BUGBOT} unknown error while retrying persisted request. Not retrying.`, {
+                    command: request.command,
+                    error: error.message,
+                });
+                PersistedRequests.remove(request);
+            }
+        }));
+
+    // Do a recursive call in case the queue is not empty after processing the current batch
+    return Promise.all(tasks)
+        .then(process);
+}
+
+function flush() {
+    if (persistedRequestsQueueRunning) {
+        return;
+    }
+
+    // ONYXKEYS.PERSISTED_REQUESTS is shared across clients, thus every client/tab will have a copy
+    // It is very important to only process the queue from leader client otherwise requests will be duplicated.
+    if (!ActiveClientManager.isClientTheLeader()) {
+        return;
+    }
+
+    persistedRequestsQueueRunning = true;
+
+    // Ensure persistedRequests are read from storage before proceeding with the queue
+    const connectionId = Onyx.connect({
```

Comment:
Sorry, it's probably just a me problem, but I'm confused how this is working and have a few questions:

1. Why connect and disconnect instead of just relying on `Onyx.connect` to keep the persisted requests up-to-date? This feels like a hack b/c we're simulating `Onyx.get` here, though I'm sure it's a necessary one for now.
1. What triggers the `PersistedRequestsQueue` to flush? The only usage I see is [here](https://github.com/Expensify/App/blob/e16211f1f91de5e585dff79270f4fef538645c69/src/libs/Network/index.js#L163-L168), and that seems like it would only run on app init?

Reply:
You're right. It's a hack that ensures we don't start processing the queue until the requests have been read from storage. This mostly only matters on app init so we could probably improve this further (as we are subscribing in `PersistedRequests` as well and could use that). We might be able to use the `createOnReadyTask` to workaround this for now (but I'm not gonna add that here).

>What triggers the PersistedRequestsQueue to flush? The only usage I see is [here](https://github.com/Expensify/App/blob/e16211f1f91de5e585dff79270f4fef538645c69/src/libs/Network/index.js#L163-L168), and that seems like it would only run on app init?

Main usage here and runs when we go from offline to online

https://github.com/Expensify/App/blob/f4c1cb2accb2cd7f7a61c3375c8fea7e7589c259/src/libs/Network/PersistedRequestsQueue.js#L90-L91